src/Corax/Querying/Matches/SortingMatches/HeapSorter.cs

```diff
@@ -0,0 +1,636 @@
+                ? sorter.SecondaryComparer.Compare(posA, posB)
+                : cmp;
+        }
+
+        static int Descending(ref NumericalMaxHeapSorter<TTermType, TSecondaryCmp> sorter, TTermType termA, int posA, TTermType termB, int posB)
+        {
+            var cmp = termB.CompareTo(termA);
+            return cmp == 0
+                ? sorter.SecondaryComparer.Compare(posA, posB)
+                : cmp;
+        }
+
+        var sorter = new NumericalMaxHeapSorter<TTermType, TSecondaryCmp>();
+        sorter.Init(documents, terms, null, descending, descending ? &Descending : &Ascending, secondaryCmp);
+        return sorter;
+    }
+
+    internal struct SkipSecondaryComparer : IComparer<int>
+    {
+        public int Compare(int x, int y)
+        {
+            throw new NotImplementedException("Used as marker for generics. Should never ever be called!");
+        }
+    }
+}
+
+internal unsafe ref struct NumericalMaxHeapSorter<TTermType, TSecondaryComparer> where TSecondaryComparer : IComparer<int>
+    where TTermType : unmanaged
+{
+    private Span<int> _documents;
+    private Span<TTermType> _terms;
+    private int _heapSize;
+    private int _heapCapacity; 
+    public bool IsDescending;
+    public TSecondaryComparer SecondaryComparer;
+
+
+    // this, termA, posA, termB, posB
+    private delegate*<ref NumericalMaxHeapSorter<TTermType, TSecondaryComparer>, TTermType, int, TTermType, int, int> _compare;
+
+    public void Init(Span<int> documents, Span<TTermType> terms, ByteStringContext allocator, bool descending,
+        delegate*<ref NumericalMaxHeapSorter<TTermType, TSecondaryComparer>, TTermType, int, TTermType, int, int> compare, TSecondaryComparer secondaryCmp)
+    {
+        IsDescending = descending;
+        _documents = documents;
+        _terms = terms;
+        _heapCapacity = documents.Length;
+        _compare = compare;
+        SecondaryComparer = secondaryCmp;
+    }
+
+    public void Insert(int document, TTermType newTerm)
+    {
+        if (_heapSize < _heapCapacity)
+        {
+            _documents[_heapSize] = document;
+            _terms[_heapSize] = newTerm;
+            _heapSize++;
+            HeapIncreaseKey(_heapSize - 1);
+            return;
+        }
+
+
+        // Since we're gathering the smallest N keys, we can do a simple comparison cmp(Max_Heap, New),
+        // as every new item will have to be smaller than the maximum in the heap.
+        int isNewTermSmallerThanCurrentMax = Compare(_terms[0], _documents[0], newTerm, document);
+
+        if (isNewTermSmallerThanCurrentMax > 0)
+        {
+            ReplaceMax(newTerm, document);
+        }
+    }
+
+    private int Compare(TTermType xDoc, int xIndex, TTermType yDoc, int yIndex)
+    {
+        var isNewTermSmallerThanCurrentMax = _compare(ref this, xDoc, xIndex, yDoc, yIndex);
+        if (isNewTermSmallerThanCurrentMax == 0)
+        {
+            isNewTermSmallerThanCurrentMax = xIndex - yIndex;
+        }
```

Comment:
aren't you already checking `posA` and `posB` in the compare?

Reply:
posA` and `posB` are used by secondary comparators to retrieve the id of documents (from batchResults) and load terms for them to compare.